content/blog/2025-05-13-polyhedral.md

```diff
@@ -0,0 +1,190 @@
+   |  instr_1; ...; instr_N; c
+   |  if b then c else c
+   |  jmp .l
+   |  ret v
+```
+
+A jump to label `.l` is only allowed when `.l` is in scope (this is what makes
+it structured!). Jumping to a `block` label is like a `break`, escaping to the
+end of the block, while jumping to a `loop` label is like a `continue`, going
+back to the loop header for the next iteration.
+
+I mostly follow the algorithm from [Beyond Relooper (PLDI'22 functional
+pearl)][beyond-relooper], though my structured IR looks a little different, and
+I don't make any effort to do anything with irreducible CFGs.
+
+[beyond-relooper]: https://dl.acm.org/doi/abs/10.1145/3547621
+
+This is far more structured than a sea of basic blocks, but still a far cry
+from our nice static control flow!  The next step is to recursively process the
+structured IR, eventually getting to the simplified polyhedral representation.
+This step is by far the most complex, accounting for a large portion of the
+code, and with many tricky cases to consider.  The biggest difference between
+the structured IR and the polyhedral IR, other than that the polyhedral IR has
+static control flow, is that code in the structured IR can have multiple exits:
+you can branch to any label that's in scope, depending on whatever conditions
+you want. By contrast, the polyhedral IR can only go on to the next statement.
+The combination of both multiple exits and recovering static control flow makes
+this a complex implementation task.
+
+[mlir]: https://mlir.llvm.org/docs/Rationale/RationaleSimplifiedPolyhedralForm/
+
+## Evaluation
+
+I evaluated the code in three different ways:
+
+ 1. Correctness
+ 2. Performance overhead incurred by round-tripping through the IR
+ 3. Applicability of potential polyhedral optimizations
+
+For all three, I used both the core Bril benchmark suite and the memory
+benchmark suite.  Sadly, as I didn't get to implement any real polyhedral
+optimizations, I don't have speed-ups to measure (yet?).
+
+**Correctness.** I tested correctness of round-tripping just through the
+relooper, and then also through the polyhedral IR, by comparing against
+interpreting the original code.
+
+**Performance.** On the combined mem and core benchmarks, round-tripping
+through the polyhedral IR gives a mean 0.78x speedup. However, my compiler at
+the moment does zero backend optimizations, which would be quite useful to
+clean up the code. I expect this to go back to close to 1x with simple backend
+clean-up optimizations.
+
+**Applicability.** Not every program has static control flow, so only some
+could potentially benefit from future polyhedral optimizations. In the combined
+mem and core benchmarks, 176 out of 272 functions had fully static control
+flow, or around 65%. Considering just the relooper, interestingly, all 272
+functions in these combined test suites exhibit reducible control flow, and
+successfully round-trip through just the structured control flow IR.
```

Comment:
I agree that this is interesting! I would have expected _some_ benchmark in there to do something weird with its control flow.